components/ExpertiseSelector.js

```diff
@@ -160,112 +35,330 @@ export default function ExpertiseSelector({ invitation, venueId, apiVersion, sho
-        setUserNotes(publicNotes)
-      } catch (error) {
-        promptError(error.message)
-        setUserNotes([])
-      }
-    }
-
-    const loadEdges = async () => {
-      try {
-        const edges = await api.getAll(
-          '/edges',
-          {
-            invitation: invitation.id,
-            tail: user.profile.id,
-          },
-          { accessToken, version: apiVersion }
+        setIncludedKeyphrases(
+          Object.keys(jsonResponse['profilekp2selected']).filter(
+            (kp) => jsonResponse['profilekp2selected'][kp]
+          )
         )
-        setEdgesMap(keyBy(edges, 'head'))
+        setRecommendedPapers(jsonResponse['recommendations'])
+        setActivePapers(jsonResponse['profilekp2user_papers'])
       } catch (error) {
-        promptError(error.message)
-        setEdgesMap({})
+        console.log('Error fetching', error)
       }
     }
 
-    loadNotes()
-    loadEdges()
-  }, [userLoading, user, shouldReload])
+    fetchKeyphrases()
+  }, [])
 
   useEffect(() => {
-    // Mark task as complete if the invitation is an "Exclude" type and it's the
-    // first time the user is visiting the page
-    if (
-      invitationOption === 'Exclude' &&
-      userNotes &&
-      edgesMap &&
-      Object.keys(edgesMap).length === 0
-    ) {
-      markTaskAsComplete()
+    function buildPaperWeightDict(jsonResponse) {
+      var paperWeightDict = {}
+      Object.keys(jsonResponse['profilekp2user_papers']).forEach((kp) => {
+        jsonResponse['profilekp2user_papers'][kp]['papers'].forEach((paper, index) => {
+          if (paperWeightDict.hasOwnProperty(paper['title'])) {
+            paperWeightDict[paper['title']] = [
+              ...paperWeightDict[paper['title']],
+              { [kp]: jsonResponse['profilekp2user_papers'][kp]['kp2paper_wts'][index] },
+            ]
+          } else {
+            paperWeightDict[paper['title']] = [
+              { [kp]: jsonResponse['profilekp2user_papers'][kp]['kp2paper_wts'][index] },
+            ]
+          }
+        })
+      })
+
+      return paperWeightDict
+    }
+    async function fetchRecommendations() {
+      const baseUrl =
+        'https://retrievalapp-zso5o2q47q-uc.a.run.app/get_recommendations/mccallum'
+      // prepend selected_kps= to every kp in includedKeyphrases
+      const url =
+        includedKeyphrases.length == 0
+          ? baseUrl
+          : baseUrl + '?selected_kps=' + includedKeyphrases.join('&selected_kps=')
+      const keyphraseResponse = await fetch(url)
+      const jsonResponse = await keyphraseResponse.json()
+      setRecommendedPapers(jsonResponse['recommendations'])
+      setActivePapers(jsonResponse['profilekp2user_papers'])
+      // in the future, paper weights should be set based on id instead of titles
+      setPaperWeights(buildPaperWeightDict(jsonResponse))
+    }
+    fetchRecommendations()
+  }, [includedKeyphrases])
+
+  useEffect(() => {
+    keyphrases &&
+      setKeyphraseColors(
```

Comment:
These colors could be computed when the keyphrases are loaded and stored in the same data structure. In general you want to minimize the number of depended state vars and useEffects.

Reply:
Thanks for the suggestion! Moved to the on load useEffect.